### ISSUE TITLE
HTTP(S) proxy auto-configuration via docker's client config.json file is broken/missing or ambiguous

### ISSUE THREAD

**Issue by alxlve** [docker/compose#6640]
## Description of the issue

HTTP(S) proxy auto-configuration via docker's client config.json file is broken/missing or ambiguous.

According to the following documentation :
[docs.docker.com/network/proxy](https://docs.docker.com/network/proxy)

Adding proxy configuration in `$DOCKER_CONFIG/config.json` (by default `$HOME/.docker/config.json` if not set) shall make docker client set automatically proxy environment variables in containers when building or running them.

`docker run` **does** populate proxy environment variables as expected in the container run.
`docker-compose up` **does not** populate proxy environment variables as expected in the container run.

## Context information

**Output of `docker-compose version`**

```
docker-compose version 1.18.0, build 8dd22a9
docker-py version: 2.6.1
CPython version: 3.6.6
OpenSSL version: OpenSSL 1.0.2k-fips  26 Jan 2017
```

**Output of `docker version`**

```
Client:
 Version:           18.09.4
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        d14af54266
 Built:             Wed Mar 27 18:34:51 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.4
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       d14af54
  Built:            Wed Mar 27 18:04:46 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `cat $DOCKER_CONFIG/config.json`**

```
{
  "proxies": {
    "default": {
      "httpProxy": "http://10.0.2.2:3128",
      "httpsProxy": "http://10.0.2.2:3128",
      "noProxy": "localhost,127.0.0.0/8,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16,.company.loc"
    }
  }
}
```

## Steps to reproduce the issue

1. Try proxy access by running `docker run busybox wget -O- example.com`, it works.

**Result**

```
<!doctype html>
<...> # Truncated HTML content
```

Here is a matching docker-compose file.

**Output of `docker-compose config`**

```
version: '3'
services:
  test:
    image: 'busybox:latest'
    restart: always
    hostname: 'localhost'
    command: sh -c "echo HTTP_PROXY=$$HTTP_PROXY && wget -O- example.com"
```

2. Try proxy access by running `docker-compose up`, it does not work.

**Result**

```
Attaching to busybox_test_1
test_1  | HTTP_PROXY=
test_1  | Connecting to example.com (93.184.216.34:80)
test_1  | wget: can't connect to remote host (93.184.216.34): Connection refused
busybox_test_1 exited with code 1
```

Workaround: I now populate proxy environment variables directly in the docker-compose file.

**Output of `docker-compose config`**

```
version: '3'
services:
  test:
    image: 'busybox:latest'
    restart: always
    hostname: 'localhost'
    environment:
      HTTP_PROXY: '$HTTP_PROXY'
      HTTPS_PROXY: '$HTTPS_PROXY'
      NO_PROXY: '$NO_PROXY'
      http_proxy: '$http_proxy'
      https_proxy: '$https_proxy'
      no_proxy: '$no_proxy'
    command: sh -c "echo HTTP_PROXY=$$HTTP_PROXY && wget -O- google.com"
```

`HTTP_PROXY`, `HTTPS_PROXY`, `NO_PROXY`, `http_proxy`, `https_proxy` and `no_proxy` environment variables are all set on the docker host machine.

3. Try proxy access by running `docker-compose up` with a workaround, it works.

**Result**

```
Attaching to withproxy_test_1
test_1  | HTTP_PROXY=http://10.0.2.2:3128
test_1  | Connecting to 10.0.2.2:3128 (10.0.2.2:3128)
test_1  | <!doctype html>
test_1  | <...> # Truncated HTML content
withproxy_test_1 exited with code 0
```

### Expected result

In step 2, proxy environment variables should have been set automatically with the values specified in docker's client `config.json`.

## Additional information

Issue #5601 and pull request #5706 (merged) indicate that the functionality is already implemented, however I have found no way to make this feature work neither I have found any documentation link.

### Documentation references

[docs.docker.com/compose/environment-variables](https://docs.docker.com/compose/environment-variables)

**Comment by chris-crone** [comment]
Hi @alxlve,

Thank you for the report! I see that you're using Compose 1.18 which is quite old. Can you please try with 1.24?

**Comment by alxlve** [comment]
Hi @chris-crone.

Thank you for your quick reply!

It seems that I have installed `docker-compose` from CentOS EPEL repository.
Here is the shell commands I had used to install `docker` and `docker-compose`.

```sh
sudo yum -y --enablerepo=extras install epel-release
sudo yum -y install yum-utils device-mapper-persistent-data lvm2
sudo yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo
sudo yum -y install docker-ce docker-ce-cli containerd.io
sudo yum -y install docker-compose
```

**Output of `which docker-compose`**

```text
/bin/docker-compose
```

**Output of `rpm -qf /bin/docker-compose`** # Finds package name of installed `docker-compose`.

```text
docker-compose-1.18.0-4.el7.noarch
```

As you said I do have an outdated `docker-compose` version. I didn't know it. I had only checked the `docker` version. I thought `docker-compose` was from the CentOS `docker` repository, so I had installed it with `yum` naively. The EPEL repository `docker-compose` is outdated and should not be used.

If I replace `docker-compose` EPEL package with a `pip` python package.

**`sudo yum remove docker-compose`** # Removes `docker-compose`

**`sudo yum -y install python-pip`** # Installs Python `pip`

`pip` did not parse any of the `HTTP_PROXY`, `HTTPS_PROXY`, `NO_PROXY`, `http_proxy`, `https_proxy` and `no_proxy` environment variables, it should parse them but it does not happen with the CentOS 7 repository version available. In order to make it work in my network environment I managed to configure `pip` proxy manually.

I created the file `/etc/pip.conf` with the following content:

```text
[global]
proxy = http://10.0.2.2:3128
```

**`sudo pip install docker-compose`** # Installs `docker-compose` from `pip`

In the future I will be able to update `docker-compose` with **`sudo pip install --upgrade docker-compose`**.

Now:

**Output of `which docker-compose`**

```text
docker-compose version 1.24.0, build 0aa5906
docker-py version: 3.7.2
CPython version: 2.7.5
OpenSSL version: OpenSSL 1.0.2k-fips  26 Jan 2017
```

I tried step 2 from first post, everything works ok!

It may not be useful to you, but I have described all this issue to help in case someone encounters the same mistake I did on a CentOS environment.

This is definitely not a `docker-compose` issue, sorry!
